src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -1893,6 +1887,23 @@ DatabaseShardImp::checkHistoricalPaths() const
     return true;
 }
 
+Backend::Counters const*
+DatabaseShardImp::getCounters() const
+{
+    std::shared_ptr<Shard> shard;
+    {
+        std::lock_guard lock(mutex_);
+        assert(init_);
+
+        auto const it{shards_.find(acquireIndex_)};
+        if (it == shards_.end())
+            return nullptr;
+        shard = it->second;
+    }
+
+    return shard->getCounters();
```

Comment:
I am really unfamiliar with the lifetime characteristics of the contents of `shards_`, but this doesn't look safe to me.  The `shard` local keeps the returned pointer valid only for the duration of this function.  That means the returned pointer is not guaranteed to be valid in the caller's context.

Maybe there are other guarantees keeping the pointer valid after the return?  But if so I think it makes sense to leave a comment describing what those guarantees are.

If I've spotted a real issue, one solution would be to not return `Backend::Counters` at all.  Return a `weak_ptr` to the correct `Shard` and let the final destination that needs `Counters` do the access.  That way the scope that actually needs to access `Counters` can verify that the requested `Shard` is still valid and lock it in place.

Another choice would be to return `Backend::Counters` by value.  That would take a bit more work because `Backend::Counters` contains a bunch of `atomic`s.  But if `Backend::Counters` were a template that could accept either `std::atomic<std::uint64_t>` or `std::uint64_t` then you could return the non-`atomic` form from this  method.

Just my take.  I could easily be way off base.

Reply:
Why even return a pointer instead of a copy?

Reply:
@scottschurr It's safe because `counters` always has a static lifetime (for everything bug the reporting mode casandra backend, but that's doesn't come into play here). Since it's static, the return value is valid even if `shard` goes away.

Reply:
@nbougalis returning a copy (`optional<Counters>`) seems reasonable. The struct is pretty small.

Reply:
The ```Backend::Counters``` returned is static but the issue you describe is possible should that change. The Shard class is a private implementation not meant to be visible beyond DatabaseShard. Since Counters was added solely for and used only by ```CassandraBackend```, I propose we have a default base implementation that returns nullptr.

Reply:
Pushed a commit with @scottschurr's template suggestion.  @scottschurr Thanks for the idea and help.